File: Python/neural.py
import cv2
import numpy as np
from PIL import Image
from mnist import MNIST
from os.path import exists
from json import load, dump

trained_data_dir = "./data/trained_data.json"
train_MNIST_data_dir = "./data/MNIST_train_data/"
color_part = 0.45
frames_multiplier = 1.5


class Neural:
    def __init__(self):
        self.coefficients = np.ones((10, 784), np.float32)

    def analyze(self, image_dir: str):
        image = self.__preprocessor(image_dir)
        results = np.zeros(10, np.float32)
        for i in range(len(results)):
            results[i] = np.sum(np.dot(self.coefficients[i], image))
        return np.argmax(results)

    def train(self, retrain=False):
        if not retrain and exists(trained_data_dir):
            self.coefficients = np.array(load(open(trained_data_dir, 'r', encoding="UTF-8")), np.float32)
            return
        train_images, train_labels = MNIST(train_MNIST_data_dir).load_training()
        train_images = self.__normalize(train_images)
        for i in range(len(train_labels)):
            image = train_images[i]
            label = train_labels[i]
            prediction = self.__analyze_train_image(image)
            if prediction != label:
                self.coefficients[label] += image
                self.coefficients[prediction] -= image
        dump(self.coefficients.tolist(), open(trained_data_dir, 'w', encoding="UTF-8"))

    def __analyze_train_image(self, image: np.array):
        results = np.zeros(10, np.float32)
        for i in range(len(results)):
            results[i] = np.sum(np.dot(self.coefficients[i], image))
        return np.argmax(results)

    def __preprocessor(self, image_dir: str):
        image = Image.open(image_dir).convert("L")
        image = image.point(self.__black_white_filter(np.asarray(image), image_dir[-3:].lower() == "jpg"))
        return self.__convert_to_square(np.asarray(image))

    def __convert_to_square(self, image: np.array):
        image = self.__crop(image)
        max_side_size = int(max(len(image), len(image[0])) * frames_multiplier)

        vertical_indent = (max_side_size - len(image)) // 2  # Находим вертикальный отступ
        horizontal_indent = (max_side_size - len(image[0])) // 2  # Находим горизонтальный отступ

        new_image = np.zeros((max_side_size, max_side_size), np.float32)
        for i in range(len(image)):
            for j in range(len(image[i])):
                new_image[i + vertical_indent][j + horizontal_indent] = image[i][j]
        return cv2.resize(new_image, (28, 28)).ravel()

    def __black_white_filter(self, image: np.array, jpg: bool):
        if jpg:
            threshold = 255 - self.__avg_color(image)

            def color_gen(x):
                return 1 if x < threshold else 0

            return color_gen
        else:
            threshold = self.__avg_color(image)

            def color_gen(x):
                return 0 if x < threshold else 1

            return color_gen

    def __crop(self, image: np.array):
        crop_coordinates = [None, None, None, None]  # y1, x1, y2, x2
        img_width = len(image[0]) - 1
        img_height = len(image) - 1
        for i in range(len(image[0])):
            if crop_coordinates[1] is None and self.__check_column(image, i):
                crop_coordinates[1] = i
            if crop_coordinates[3] is None and self.__check_column(image, img_width - i):
                crop_coordinates[3] = img_width - i
            if crop_coordinates[1] is not None and crop_coordinates[3] is not None:
                break
        for i in range(len(image)):
            if crop_coordinates[0] is None and self.__check_row(image, i):
                crop_coordinates[0] = i
            if crop_coordinates[2] is None and self.__check_row(image, img_height - i):
                crop_coordinates[2] = img_height - i
            if crop_coordinates[0] is not None and crop_coordinates[2] is not None:
                break
        return image[crop_coordinates[0]:crop_coordinates[2], crop_coordinates[1]: crop_coordinates[3]]

    @staticmethod
    def __check_row(matrix: np.array, index: int):
        for i in range(len(matrix[index])):
            if matrix[index][i] == 1:
                return True
        return False

    @staticmethod
    def __check_column(matrix: np.array, index: int):
        for i in range(len(matrix)):
            if matrix[i][index] == 1:
                return True
        return False

    @staticmethod
    def __normalize(array: list):
        for i in range(len(array)):
            array[i] = np.array(array[i]) / 255
        return array

    @staticmethod
    def __avg_color(image: np.array):
        return int(np.sum(image) / (len(image) * len(image[0])) * color_part) + 1


if __name__ == "__main__":
    neural = Neural()
    neural.train()

    samples_dir = "./samples/"
    samples_list = ["eleven.jpg", "four.png", "one_1.png", "one_2.png", "seven_1.png", "seven_2.jpg", "six_1.png",
                    "six_2.jpg", "six_3.jpg"]
    for sample in samples_list:
        print(neural.analyze(samples_dir + sample))

File: Java/src/com/neural/network/Neural.java
package com.neural.network;

import com.google.gson.Gson;

import javax.imageio.ImageIO;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.*;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;

public class Neural {

    private final double colorPart = 0.75d;
    private final double frameMultiplier = 1.5d;

    private final String trained_data = "src/com/neural/newnetwork/data/trained_data.json";
    private final String MNIST_data = "src/com/neural/newnetwork/data/MNIST_data/";

    private int width;  // Сюда будем сохранять ширину картинки
    private int height;  // Сюда будем сохранять высоту картинки
    private double[][] coefficients;

    public Neural(){
        coefficients = new double[10][784];
    }

    public int analyze(String imgDir){
        int[] image = preprocessor(imgDir);
        double[] results = new double[10];
        for (int i = 0; i < results.length; i++){
            results[i] = Linear.dotAndSum(coefficients[i], image);
        }
        return Linear.argMax(results);
    }

    public int analyze(double[] image){
        double[] results = new double[10];
        for (int i = 0; i < results.length; i++){
            results[i] = Linear.dotAndSum(coefficients[i], image);
        }
        return Linear.argMax(results);
    }

    public void train(boolean retrain){
        if (!retrain && Files.exists(Paths.get(trained_data))){
            try {
                coefficients = new Gson().fromJson(new FileReader(new File(trained_data)), double[][].class);
            } catch (FileNotFoundException e) {
                e.printStackTrace();
            }
        } else {
            double[][] images = normalize(MnistReader.getImages(MNIST_data + "train_images"));
            int[] labels = MnistReader.getLabels(MNIST_data + "train_labels");
            for (int i = 0; i < labels.length; i++){
                double[] image = images[i];
                int label = labels[i];
                int prediction = analyze(image);
                if (prediction != label){
                    coefficients[prediction] = Linear.minus(coefficients[prediction], image);
                    coefficients[label] = Linear.plus(coefficients[label], image);
                }
            }
            try{
                FileWriter writer = new FileWriter(trained_data);
                writer.write(new Gson().toJson(new Gson().toJsonTree(coefficients)));
                writer.close();
            } catch (IOException e){
                e.printStackTrace();
            }
        }
    }

    private int[] preprocessor(String imgDir){
        int[] result = null;
        try{
            BufferedImage image = ImageIO.read(new File(imgDir));
            result = toSquare(blackWhiteFilter(grayScale(image), imgDir.substring(imgDir.length() - 3).toLowerCase().equals("jpg")));
        } catch (IOException e){
            e.printStackTrace();
        }
        return result;
    }

    private int[] toSquare(int[] image) {
        int[][] cropped = crop(image);
        int max_size_side = (int) (Math.max(cropped.length, cropped[0].length) * frameMultiplier);
        int vertical_indent = (max_size_side - cropped.length) / 2;
        int horizontal_indent = (max_size_side - cropped[0].length) / 2;
        int[][] new_image = new int[max_size_side][max_size_side];
        for (int i = 0; i < cropped.length; i++){
            for (int j = 0; j < cropped[0].length; j++){
                new_image[i + vertical_indent][j + horizontal_indent] = cropped[i][j];
            }
        }
        width = max_size_side;
        height = max_size_side;
        return resize(Linear.ravel(new_image), 28, 28);
    }

    private int[] resize(int[] pixels, int width, int height) {
        BufferedImage image = new BufferedImage(this.width, this.height, BufferedImage.TYPE_INT_RGB);
        for (int y = 0; y < this.height; y++){
            for (int x = 0; x < this.width; x++){
                image.setRGB(x, y, pixels[y * this.width + x]);
            }
        }

        BufferedImage scaledBI = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
        Graphics2D g = scaledBI.createGraphics();
        g.drawImage(image, 0, 0, width, height, null);
        g.dispose();

        int[] result = new int[width * height];
        for (int i = 0; i < height; i++){
            for (int j = 0; j < width; j++){
                result[i * width + j] = scaledBI.getRGB(j, i) & 0xff;
            }
        }
        return result;
    }

    private int[][] crop(int[] image) {
        int[] coordinates = new int[]{0, 0, 0, 0};  // y1, x1, y2, x2
        for (int i = 0; i < image.length; i += width){
            for (int j = i; j < i + width; j++){
                coordinates[0] = Math.min(coordinates[0], i / width);
                coordinates[1] = Math.min(coordinates[1], j - i);
                coordinates[2] = Math.max(coordinates[2], i / width);
                coordinates[3] = Math.max(coordinates[3], j - i);
            }
        }
        return cutter(reshape(image, width, height), coordinates);
    }

    private int[][] cutter(int[][] image, int[] coordinates){
        int[][] result = new int[coordinates[2] - coordinates[0] + 1][coordinates[3] - coordinates[1] + 1];
        for (int i = 0; i < image.length; i++){
            for (int j = 0; j < image[i].length; j++){
                if (i <= coordinates[2] && i >= coordinates[0]  && j <= coordinates[3] && j >= coordinates[1]){
                    result[i - coordinates[0]][j - coordinates[1]] = image[i][j];
                }
            }
        }
        return result;
    }

    private int[][] reshape(int[] array, int width, int height){
        int[][] result = new int[height][width];
        for (int i = 0; i < array.length / width; i++){
            for (int j = 0; j < width; j++){
                result[i][j] = array[i * width + j];
            }
        }
        return result;
    }

    private int[] blackWhiteFilter(int[] array, boolean jpg){
        int avgColor = avgColor(array);
        int[] result = new int[array.length];
        for (int i = 0; i < array.length; i++){
            if (array[i] >= avgColor){
                result[i] = jpg ? 0: 1;
            } else {
                result[i] = jpg ? 1: 0;
            }
        }
        return result;
    }

    private int avgColor(int[] array) {
        return (int) (Linear.sum(array) / array.length * colorPart) + 1;
    }

    private int[] grayScale(BufferedImage image){
        width = image.getWidth();
        height = image.getHeight();
        int[] result = new int[width * height];
        for (int y = 0; y < height; y++){
            for (int x = 0; x < width; x++){
                int pixel = image.getRGB(x, y);
                int r = (pixel >> 16) & 0xff;
                int g = (pixel >> 8) & 0xff;
                int b = pixel & 0xff;
                result[y * width + x] = (int) Math.ceil((r + g + b) / 3);
            }
        }
        return result;
    }

    private double[][] normalize(List<int[][]> images){
        double[][] result = new double[images.size()][images.get(0).length * images.get(0)[0].length];
        for (int i = 0; i < images.size(); i++){
            result[i] = Linear.div(Linear.ravel(images.get(i)), 255d);
        }
        return result;
    }

    public static void main(String[] args) {
        Neural neural = new Neural();
        neural.train(false);
        List<String> samples = List.of("eleven.jpg", "four.png", "one_1.png", "one_2.png",
                "seven_1.png", "seven_2.jpg", "six_1.png", "six_2.jpg", "six_3.jpg");
        for (String sample: samples){
            //if (sample.substring(sample.length() - 3).equals("jpg")) continue;
            System.out.println(sample);
            System.out.println(neural.analyze("src/com/neural/newnetwork/samples/" + sample));
            //if (sample.equals("one_1.png")) System.exit(0);
        }
    }
}
